frontend/src/APIClients/types/CourseClientTypes.ts

```diff
@@ -0,0 +1,99 @@
+type Module = {
+  title: string;
+  description: string;
+  image: string;
+  previewImage: string;
+  published: boolean;
+  lessons: string[];
+};
+
+type SerializedModule = Module & { id: string };
+
+type ModulesById = { [id: string]: Module };
+
+export type CourseRequest = {
+  title: string;
+  description: string;
+  image: string;
+  previewImage: string;
+  modules: ModulesById;
+  private: boolean;
+  published: boolean;
+};
+
+export type CourseResponse = {
+  id: string;
+  title: string;
+  description: string;
+  image: string;
+  previewImage: string;
+  modules: ModulesById;
+  private: boolean;
+  published: boolean;
+};
+
+type SerializedCourseResponse = Omit<CourseResponse, "modules"> & {
+  modules: SerializedModule[];
+};
+type SerializedCreateCourseRequest = Omit<CourseRequest, "modules"> & {
+  modules: SerializedModule[];
+};
+type SerializedUpdateCourseRequest = Partial<SerializedCreateCourseRequest>;
+
+// Helper functions
+export const serializeModules = (modules: ModulesById): SerializedModule[] => {
+  return Object.entries(modules).map(([module_id, module]) => ({
+    id: module_id,
+    ...module,
+  }));
+};
+
+export const deserializeModules = (
+  serModules: SerializedModule[],
+): ModulesById => {
+  const modules: { [id: string]: Module } = {} as { [id: string]: Module };
+
+  serModules.forEach((module) => {
+    const moduleObj = {
+      title: module.title,
+      description: module.description,
+      image: module.image,
+      previewImage: module.previewImage,
+      published: module.published,
+      lessons: module.lessons,
+    };
+
+    modules[module.id] = moduleObj;
+  });
+
+  return modules;
+};
+
+export const serializeCreateCourseRequest = (
+  courseRequest: CourseRequest,
+): SerializedCreateCourseRequest => {
+  return {
+    ...courseRequest,
+    modules: serializeModules(courseRequest.modules),
+  };
+};
+
+export const serializeUpdateCourseRequest = (
+  courseRequest: Partial<CourseRequest>,
+): SerializedUpdateCourseRequest => {
+  return {
+    modules:
+      courseRequest.modules === undefined
+        ? undefined
+        : serializeModules(courseRequest.modules),
```

Comment:
How does this handle re: typing? As in, code pathways seem to have a "return undefined" path but the method signature does not

Reply:
Hm,, this is actually extracted from the [implement Modules PR](https://github.com/uwblueprint/rowan-house/pull/92/files#diff-7c71de3be3e3b0b16555de131be1a187b67a7b94367a1da8b9721cb98d1cb68aR84-R93). I'm just reorganizing these helper functions into a separate file to be accessed when the course gql strings are used.
@Shrinjay could you address this comment?